contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,231 @@
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external returns (bytes32) {
-        require(msg.sender == admin, "Timelock::queueTransaction: Call must come from admin.");
-        require(eta >= getBlockTimestamp() + delay, "Timelock::queueTransaction: Estimated execution block must satisfy delay.");
+    function approveTransactionsBatch(bytes32 txParamsHash) external onlyOwner {
+        _approveTransactionsBatchInternal(txParamsHash);
+    }
+
+    function executeTransactionsBatch(Transaction[] memory transactions, uint256 eta) external payable onlyExecutorOrOwner {
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
+
+    function approveAndExecuteTransactionsBatch(Transaction[] memory transactions, uint256 eta) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, eta));
+
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
 
-        bytes32 txHash = keccak256(abi.encode(target, value, signature, data, eta));
-        queuedTransactions[txHash] = true;
+    function approveAndExecuteTransactionsBatchRaw(bytes calldata txParamsData) external payable onlyOwner {
+        bytes32 txParamsHash = keccak256(txParamsData);
+        (Transaction[] memory transactions, uint256 eta) = abi.decode(txParamsData, (Transaction[], uint256));
 
-        emit QueueTransaction(txHash, target, value, signature, data, eta);
-        return txHash;
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, eta);
     }
 
-    function cancelTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external {
-        require(msg.sender == admin, "Timelock::cancelTransaction: Call must come from admin.");
+    function _approveTransactionsBatchInternal(bytes32 txParamsHash) internal onlyOwner {
+        require(pendingTransactions[txParamsHash] == TransactionStatus.Queued, "Timelock: transaction is not queued");
 
-        bytes32 txHash = keccak256(abi.encode(target, value, signature, data, eta));
-        queuedTransactions[txHash] = false;
+        pendingTransactions[txParamsHash] = TransactionStatus.Approved;
 
-        emit CancelTransaction(txHash, target, value, signature, data, eta);
+        emit ApprovedTransactionsBatch(txParamsHash);
     }
 
-    function executeTransaction(
-        address target,
-        uint256 value,
-        string memory signature,
-        bytes memory data,
-        uint256 eta
-    ) external payable returns (bytes memory) {
-        require(msg.sender == admin, "Timelock::executeTransaction: Call must come from admin.");
+    function _executeTransactionsBatchInternal(Transaction[] memory transactions, uint256 eta) internal nonReentrant {
```

Comment:
Shouldn't that be `private` instead of `internal`?

> internal :Those functions and state variables can only be accessed internally (i.e. from within the current contract or contracts deriving from it), without using this . private :Private functions and state variables are only visible for the contract they are defined in and not in derived contracts.

That also applies to `_approveTransactionsBatchInternal`

Reply:
ok